src/ethereum_spec_tools/diff.py

```diff
@@ -0,0 +1,161 @@
+#!/usr/bin/env python3
+
+"""
+Generates diffs between Ethereum hardforks documentation.
+"""
+
+import os.path
+import pickle
+from typing import Iterator, List, TypeVar
+
+import rstdiff
+from docutils import SettingsSpec
+from docutils.utils import new_document, new_reporter
+
+from ethereum_spec_tools.forks import Hardfork
+
+T = TypeVar("T")
+
+
+def window(forks: List[T], window_size: int = 2) -> Iterator[List[T]]:
+    """
+    Group a list into overlapping chunks.
+    """
+    for i in range(len(forks) - window_size + 1):
+        yield forks[i : i + window_size]
+
+
+def find_pickles(path: str, fork: Hardfork) -> Iterator[str]:
+    """
+    Find files ending with :code:`.pickle`.
+    """
+    for directory, _, filenames in os.walk(path):
+        for filename in filenames:
+            if filename.endswith(".pickle"):
+                file_path = os.path.join(directory, filename)
+                file_path = file_path[len(path) + 1 :]
+                yield file_path
+
+
+def diff(
+    output_path: str, input_path: str, old: Hardfork, new: Hardfork
+) -> Iterator[str]:
+    """
+    Calculate the structured diff between two hardforks.
+    """
+    old_path = old.name.replace(".", os.sep)
+    old_path = os.path.join(input_path, old_path)
+
+    new_path = new.name.replace(".", os.sep)
+    new_path = os.path.join(input_path, new_path)
+
+    diff_path = old.name.split(".")[-1] + "_" + new.name.split(".")[-1]
+    diff_path = os.path.join(output_path, diff_path)
+
+    old_pickles = set(find_pickles(old_path, old))
+    new_pickles = set(find_pickles(new_path, new))
+
+    pickles = old_pickles | new_pickles
+
+    for count, pickle_path in enumerate(pickles, start=1):
+        old_pickle_path = os.path.join(old_path, pickle_path)
+        new_pickle_path = os.path.join(new_path, pickle_path)
+        diff_pickle_path = os.path.join(diff_path, pickle_path + "64")
```

Comment:
Why `64`?

Reply:
The build process has two stages, `stage0` and `stage1`.

`stage0` writes to binary pickle files, and those files are opened by `rstdiff`, which is fine handling binary files.

Sphinx, on the other hand, only lets you define parsers in extensions and not readers. Parsers can only handle strings (not `bytes`) so the input has to be valid UTF-8.

The output from the `rstdiff` is written to `.pickle64` files (base64 encoded pickle) to satisfy that requirement.

Reply:
ahh okay gotcha!